### PR TITLE
Prevent traceback chaining in _wrapfunc.

### PR DESCRIPTION
**Pull request by anntzer**
The traceback of `np.reshape([1, 2, 3], 2)` is shortened from
```
Traceback (most recent call last):
  File ".../numpy/core/fromnumeric.py", line 56, in _wrapfunc
    return getattr(obj, method)(*args, **kwds)
AttributeError: 'list' object has no attribute 'reshape'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../numpy/core/overrides.py", line 151, in public_api
    implementation, public_api, relevant_args, args, kwargs)
  File ".../numpy/core/fromnumeric.py", line 296, in reshape
    return _wrapfunc(a, 'reshape', newshape, order=order)
  File ".../numpy/core/fromnumeric.py", line 66, in _wrapfunc
    return _wrapit(obj, method, *args, **kwds)
  File ".../numpy/core/fromnumeric.py", line 46, in _wrapit
    result = getattr(asarray(obj), method)(*args, **kwds)
ValueError: cannot reshape array of size 3 into shape (2,)
```
to
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../numpy/core/overrides.py", line 151, in public_api
    implementation, public_api, relevant_args, args, kwargs)
  File ".../numpy/core/fromnumeric.py", line 300, in reshape
    return _wrapfunc(a, 'reshape', newshape, order=order)
  File ".../numpy/core/fromnumeric.py", line 70, in _wrapfunc
    return _wrapit(obj, method, *args, **kwds)
  File ".../numpy/core/fromnumeric.py", line 46, in _wrapit
    result = getattr(asarray(obj), method)(*args, **kwds)
ValueError: cannot reshape array of size 3 into shape (2,)
```

(The chained exception is really just an implementation detail.)

### CODE REVIEW

**Review by mhvk — COMMENTED** [review]
I think this is a good idea for the `AttributeError`, but perhaps less so for the `TypeError`, where it is more likely something is passed in that should just to be dealt with anyway.

**Comment by anntzer** [comment]
That would give something like
```python
def _wrapfunc(obj, method, *args, **kwds):
    try:
        bound = getattr(obj, method)
    except AttributeError:
        pass
        # An AttributeError occurs if the object does not have such a method in
        # its class.
    else:
        try:
            return bound(*args, **kwds)
        except TypeError:
            # A TypeError occurs if the object does have such a method in
            # its class, but its signature is not identical to that of
            # NumPy's. This situation has occurred in the case of a downstream
            # library like 'pandas'.
            #
            # In this case, we want to traceback-chain a potential exception.
            return _wrapit(obj, method, *args, **kwds)
    # When the method doesn't exist in the object's class, don't do the
    # fallback in the except handler, to avoid traceback chaining in case it
    # also raises an exception.
    return _wrapit(obj, method, *args, **kwds)
```
Is the complexity worth it? (perhaps, dunno, just asking)

**Comment by mhvk** [comment]
@anntzer - I also don't quite know...  As I indicated, at some level I don't understand why we catch the `TypeError` at all - is it actually useful for a method to fail and then turn the object in an array and do it again? 

Though that's perhaps beyond this nice & simple PR! For this PR, my sense is that it is worth it, and I'd perhaps add a `TODO: should we just raise?` to the comment for the `TypeError`.

But maybe we can see what others think...

**Comment by anntzer** [comment]
I don't really want to get into the discussion of whether TypeErrors should be caught or not; I'll leave it to the numpy devs to decide :) and update the patch accordingly if needed.

**Comment by mhvk** [comment]
Indeed, let's just stick with the chaining here.  Let me ping @eric-wieser for an independent opinion.

**Comment by eric-wieser** [comment]
I suspect that someone would be broken if we stopped silencing `TypeError`s and coercing to arrays, so I think we need to continue to catch them.

The nesting becomes a little clearer as:
```
def _wrapfunc(obj, method, *args, **kwds):
    bound = getattr(obj, method, None)
    if not bound:
        return _wrapit(obj, method, *args, **kwds)

    try:
        return bound(*args, **kwds)
    except TypeError:
        # A TypeError occurs if the object does have such a method in
        # its class, but its signature is not identical to that of
        # NumPy's. This situation has occurred in the case of a downstream
        # library like 'pandas'.
        #
        # In this case, we want to traceback-chain a potential exception.
        return _wrapit(obj, method, *args, **kwds)
```

**Comment by anntzer** [comment]
@eric-wieser's solution looks good, pushed it.

**Review by mhvk — APPROVED** [review]
I like this solution very much!

File: numpy/core/fromnumeric.py
Comment (mhvk) [inline numpy/core/fromnumeric.py]: I noticed that the commit message doesn't quite fit policy: It should have a `MAINT: ` in front.  Could you change it? (Not sure how to do it as a maintainer.)

And that is perhaps just enough reason to also rewrite the comment here: "In this case," now reads oddly. Maybe something like: "Call _wrapit from within the except clause to ensure a potential exception has a traceback chain".

With that done, I'll merge...

Comment (mattip) [reply numpy/core/fromnumeric.py]: There is an Edit button next to the PR title. Note you can also change the merge message when you press the "Merge Pull Request" button

Comment (anntzer) [reply numpy/core/fromnumeric.py]: both points handled

Comment (mhvk) [reply numpy/core/fromnumeric.py]: @mattip - that changes the PR title or the merge commit title, but not the commit with the actual work (which I think is the one we want...).

@anntzer - thanks for the update, will now merge.

**Pull request merged**
